dragonfly/implementations/at_operator.py

```diff
@@ -0,0 +1,460 @@
+        operator_name: the temporary operator to be removed; the user itself will be removedd if empty
+        '''
+        remove = user_id
+        if operator_name != "":
+            if operator_name not in self.username_to_id_dictionary:
+                self.send_message(channel, "Sorry, I don't rocognize that username.")
+                return
+            remove = self.username_to_id_dictionary[operator_name]
+        if remove not in self.temporary_operator_id:
+            self.send_message(channel, self.id_to_username_dictionary[remove] + " is not currently listed as a temporary operator.")
+            return
+        self.temporary_operator_id.remove(remove)
+        self.send_message(channel, "Ok, you're all done. Thanks!")
+
+    def construct_command_dictionary(self):
+        '''
+        Return a dictionary containing all Slack helper commands.
+        '''
+        self.command_dictionary["!hello"] = self.command_hello
+        self.command_dictionary["!help"] = self.command_help
+        self.command_dictionary["!whoisop"] = self.command_whoisop
+        self.command_dictionary["!tempoperator"] = self.command_tempoperator
+        self.command_dictionary["!removetempoperator"] = self.command_removetempoperator
+        logger.info(' Constructed dictionaries for operator helper commands on Slack.')
+
+    def parse_output(self, rtm_output, bot_id):
+        '''
+        Parse the given Slack output and check whether or not I am called.
+        rtm_output: Slack runtime output
+        bot_id    : my bot id
+        '''
+        output = rtm_output
+        if output and len(output) > 0:
+            for o in output:
+                if o and 'text' in o:
+                    if ('@' + str(bot_id)) in o['text']:
+                        return o['channel']
+                    tokens = o['text'].split(' ', 1)
+                    command = tokens[0]
+                    if command in self.command_dictionary:
+                        logger.info(' A helper command [' + command + '] is called!')
+                        channel = o["channel"]
+                        user_id = o["user"]
+                        operator_name = ""
+                        if len(tokens) > 1 and tokens[1].startswith('[') and "]" in tokens[1]:
+                            sub_tokens = tokens[1].split(']', 1)
+                            operator_name = sub_tokens[0].replace('[', '')
+                            logger.info(' Found a potential operator name [' + operator_name + '] !')
+                        func = self.command_dictionary[command]
+                        num_args = len(funcsigs.signature(func).parameters)
+                        args = [channel, user_id, operator_name]
+                        func(*args[:num_args])
+        return None
+
+    def run(self):
+        '''
+        The main loop.
+        '''
+        logging.getLogger('googleapiclient.discovery_cache').setLevel(logging.ERROR)
```

Comment:
Is this what causes the logging thread to duplicate everything?

Reply:
I don't think so - I put this line here to hide(?) an unnecessary log produced by googleapiclient, which is supposed to have no impact on the Operator logger. And removing this line doesn't stop the duplication.

And so far I have no idea why the duplication would happen :(((